In [38]:
import pandas as pd
import re



# Загрузка данных

In [39]:
data = pd.read_csv('titanic.csv', index_col='PassengerId')

In [40]:
data

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
13                  0       3   
14                  0       3   
15                  0       3   
16                  1       2   
17                  0       3   
18                  1       2   
19                  0       3   
20                  1       3   
21                  0       2   
22                  1       2   
23                  1       3   
24                  1       1   
25                  0       3   
26                  1       3   
27                  0       3   
28                  0       1   
29                  1       3   
30                  0       3   
...               ...     ...   
862                 0       2   
863                 1       1   
864                 0       3   
865                 0       2   
866                 1       2   
867                 1       2   
868                 0       1   
869                 0       3   
870                 1       3   
871                 0       3   
872                 1       1   
873                 0       1   
874                 0       3   
875                 1       2   
876                 1       3   
877                 0       3   
878                 0       3   
879                 0       3   
880                 1       1   
881                 1       2   
882                 0       3   
883                 0       3   
884                 0       2   
885                 0       3   
886                 0       3   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
13                              Saundercock, Mr. William Henry    male  20.0   
14                                 Andersson, Mr. Anders Johan    male  39.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
17                                        Rice, Master. Eugene    male   2.0   
18                                Williams, Mr. Charles Eugene    male   NaN   
19           Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0   
20                                     Masselmani, Mrs. Fatima  female   NaN   
21                                        Fynney, Mr. Joseph J    male  35.0   
22                                       Beesley, Mr. Lawrence    male  34.0   
23           

In [41]:
data.shape

(891, 11)

# 1. Какое количество мужчин и женщин ехало на корабле? В качестве ответа приведите два числа через пробел

In [42]:
data['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

**Ответ:**

577 314

# 2. Какой части пассажиров удалось выжить? Посчитайте долю выживших пассажиров. Ответ приведите в процентах (число в интервале от 0 до 100, знак процента не нужен), округлив до двух знаков

In [43]:
data['Survived'].value_counts(normalize=True) * 100

0    61.616162
1    38.383838
Name: Survived, dtype: float64

# **Ответ:**

38.38

# 3. Какую долю пассажиры первого класса составляли среди всех пассажиров? Ответ приведите в процентах (число в интервале от 0 до 100, знак процента не нужен), округлив до двух знаков

In [44]:
data['Pclass'].value_counts(normalize=True) * 100

3    55.106622
1    24.242424
2    20.650954
Name: Pclass, dtype: float64

**Ответ:**

24.24

# 4. Какого возраста были пассажиры? Посчитайте среднее и медиану возраста пассажиров. В качестве ответа приведите два числа через пробел

In [45]:
data['Age'].mean()

29.69911764705882

In [46]:
data['Age'].median()

28.0

**Ответ:**

29.70 28.00

# 5. Коррелируют ли число братьев/сестер/супругов с числом родителей/детей? Посчитайте корреляцию Пирсона между признаками SibSp и Parch

In [47]:
data[['SibSp', 'Parch']].corr()

SibSp     Parch
SibSp  1.000000  0.414838
Parch  0.414838  1.000000

**Ответ:**

0.41

# 6. Какое самое популярное женское имя на корабле? Извлеките из полного имени пассажира (колонка Name) его личное имя (First Name). Это задание — типичный пример того, с чем сталкивается специалист по анализу данных. Данные очень разнородные и шумные, но из них требуется извлечь необходимую информацию. Попробуйте вручную разобрать несколько значений столбца Name и выработать правило для извлечения имен, а также разделения их на женские и мужские

In [48]:
names = data['Name'][data['Sex'] == 'female']

In [49]:
names

PassengerId
2      Cumings, Mrs. John Bradley (Florence Briggs Th...
3                                 Heikkinen, Miss. Laina
4           Futrelle, Mrs. Jacques Heath (Lily May Peel)
9      Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
10                   Nasser, Mrs. Nicholas (Adele Achem)
11                       Sandstrom, Miss. Marguerite Rut
12                              Bonnell, Miss. Elizabeth
15                  Vestrom, Miss. Hulda Amanda Adolfina
16                      Hewlett, Mrs. (Mary D Kingcome) 
19     Vander Planke, Mrs. Julius (Emelia Maria Vande...
20                               Masselmani, Mrs. Fatima
23                           McGowan, Miss. Anna "Annie"
25                         Palsson, Miss. Torborg Danira
26     Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...
29                         O'Dwyer, Miss. Ellen "Nellie"
32        Spencer, Mrs. William Augustus (Marie Eugenie)
33                              Glynn, Miss. Mary Agatha
39                 

In [169]:
names.str.extract('((?<=Miss. )|(?<=\())(\w*)')

# names.str.extract('((?<=\()|(?<=Miss. )|(?<=Masselmani, Mrs. ))(\w*)')[1].value_counts().sum()


/Users/WLMike/anaconda3/envs/ML36/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.


0           1
PassengerId                 
2                   Florence
3                      Laina
4                       Lily
9                  Elisabeth
10                     Adele
11                Marguerite
12                 Elizabeth
15                     Hulda
16                      Mary
19                    Emelia
20           NaN         NaN
23                      Anna
25                   Torborg
26                     Selma
29                     Ellen
32                     Marie
33                      Mary
39                   Augusta
40                    Jamila
41                   Johanna
42                   Dorothy
44                   Simonne
45                  Margaret
48                   Bridget
50                  Josefine
53                      Myna
54                 Elizabeth
57                     Emily
59                 Constance
62                    Amelie
...          ...         ...
808                    Ellen
810                   Bertha
814                     Ebba
817                   Wendla
821                    Clara
824                    Beila
830                   Martha
831                   Selini
836                     Sara
843                  Augusta
850                   Edwiga
853                Nourelain
854                     Mary
855                   Lilian
856                     Leah
857                     Mary
859                   Latifa
863                 Margaret
864                  Dorothy
866                 Karolina
867                 Asuncion
872                   Sallie
875                   Hannah
876                    Adele
880                     Lily
881                  Imanita
883                    Gerda
886                 Margaret
888                 Margaret
889                Catherine

[314 rows x 2 columns]

In [171]:
names[names.str.extract('((?<=Miss. )|(?<=\())(\w*)')[0].isna()]

/Users/WLMike/anaconda3/envs/ML36/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.


PassengerId
20            Masselmani, Mrs. Fatima
257    Thorne, Mrs. Gertrude Maybelle
370     Aubart, Mme. Leontine Pauline
444         Reynaldo, Ms. Encarnacion
642              Sagesser, Mlle. Emma
707     Kelly, Mrs. Florence "Fannie"
798                  Osman, Mrs. Mara
Name: Name, dtype: object

In [179]:
names.str.extract('((?<=Miss. )|(?<=\())(\w*)')[1].value_counts()

/Users/WLMike/anaconda3/envs/ML36/lib/python3.6/site-packages/ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.


Anna          15
Mary          14
Elizabeth     11
Margaret      10
Alice          6
Maria          5
Bertha         5
Helen          5
Augusta        4
Ellen          4
Catherine      4
Annie          4
Emily          4
Marie          4
Ada            4
Florence       3
Elisabeth      3
Hanora         3
Madeleine      3
Katherine      3
Marion         3
Amelia         3
Jane           3
Lillian        3
Kate           3
Ida            3
Edith          3
Ethel          2
Charlotte      2
Jessie         2
              ..
Eliina         1
Hannah         1
Sallie         1
Margaretta     1
Imanita        1
Carla          1
Selma          1
Amelie         1
Katie          1
Phyllis        1
Ellis          1
Katriina       1
Brigdet        1
Banoura        1
Erna           1
Stella         1
Sigrid         1
Ingeborg       1
Velin          1
Ruth           1
Aurora         1
Elna           1
Elise          1
Alma           1
Farnham        1
Jamila         1
Marian         1
Grace         

**Ответ:**

Anna